sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -1061,3 +1118,149 @@ case class GpuToCpuCollectBufferTransition(
+  // which only accept one column in the input.
+  //
+  // We cast `n` to be an Integer, as that's what MERGE_M2 expects. Note that Spark keeps
+  // `n` as Double thus we also need to cast `n` back to Double after merging.
+  // In the future, we need to rewrite CudfMergeM2 such that it accepts `n` in Double type and
+  // also output `n` in Double type.
+  override lazy val preMerge: Seq[Expression] = {
+    val childrenWithNames =
+      GpuLiteral("n", StringType) :: GpuCast(bufferN, IntegerType) ::
+        GpuLiteral("avg", StringType) :: bufferAvg ::
+        GpuLiteral("m2", StringType) :: bufferM2 :: Nil
+    GpuCreateNamedStruct(childrenWithNames) :: Nil
+  }
+
+  def mergeM2DataType: DataType =
+    StructType(
+      StructField("n", IntegerType, nullable = false) ::
+        StructField("avg", DoubleType, nullable = false) ::
+        StructField("m2", DoubleType, nullable = false) :: Nil)
+
+  override def mergeBufferAttributes: Seq[AttributeReference] = postMergeAttr
+
+  private val m2Struct = AttributeReference("m2struct", mergeM2DataType, nullable = false)()
+  override lazy val mergeExpressions: Seq[Expression] = new CudfMergeM2(m2Struct) :: Nil
+
+  // The result of merging step is a structs column thus we create this attribute to bind it.
+  override lazy val postMergeAttr: Seq[AttributeReference] = Seq(m2Struct)
+
+  // The postMerge step needs to extract 3 columns (n, avg, m2) from the structs column
+  // output from the merge step. Note that the first one is casted to Double to match with Spark.
+  //
+  // In the future, when rewriting CudfMergeM2, we will need to ouput it in Double type.
+  //
+  override lazy val postMerge: Seq[Expression] = Seq(
+    GpuCast(GpuGetStructField(m2Struct, 0), DoubleType),
+    GpuCast(GpuGetStructField(m2Struct, 1), DoubleType),
+    GpuCast(GpuGetStructField(m2Struct, 2), DoubleType))
+}
+
+case class GpuStddevPop(child: Expression) extends GpuM2(child) {
+  override lazy val evaluateExpression: Expression = {
+    // stddev_pop = sqrt(m2 / n).
+    val stddevPop = GpuSqrt(GpuDivide(bufferM2, bufferN, failOnErrorOverride = false))
+
+    // Set nulls for the rows where n == 0.
+    GpuIf(GpuEqualTo(bufferN, GpuLiteral(0.0)), GpuLiteral(null, DoubleType), stddevPop)
+  }
+
+  override def children: Seq[Expression] = Seq(child)
+  override def prettyName: String = "stddev_pop"
+}
+
+case class GpuStddevSamp(child: Expression) extends GpuM2(child) {
+  override lazy val evaluateExpression: Expression = {
+    // stddev_samp = sqrt(m2 / (n - 1.0)).
+    val stddevSamp =
+      GpuSqrt(GpuDivide(bufferM2, GpuSubtract(bufferN, GpuLiteral(1.0), failOnError = false),
+        failOnErrorOverride = false))
+
+    // Set nulls for the rows where n == 0 and n == 1.
+    GpuIf(GpuEqualTo(bufferN, GpuLiteral(1.0)), GpuLiteral(null, DoubleType),
+      GpuIf(GpuEqualTo(bufferN, GpuLiteral(0.0)), GpuLiteral(null, DoubleType), stddevSamp))
```

Comment:
I just realize that Spark may set `null` or `NaN` for the case when `n === 1` (https://github.com/apache/spark/blob/adbea252db0f61ed3aaa45ea20b704bad6c47408/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala#L198). The option is set by:
```
nullOnDivideByZero: Boolean = !SQLConf.get.legacyStatisticalAggregate
```
Should we do exactly the same?

Reply:
Yes we need to do the same.

However [that change](https://issues.apache.org/jira/browse/SPARK-13860) went into Spark 3.1.0. It was a bug fix so we need to decide if we just put the change in everywhere or if we want it shimmed so we ignore that config in Spark 3.0.x but not in 3.1.0+.  I personally am fine either way.

Reply:
Let have it here instead of shim for simplicity and time sensitivity. We can always go back and work for a shim.